src/printer.rs

```diff
@@ -430,33 +431,49 @@ impl<'a> Printer for InteractivePrinter<'a> {
 
             for &(style, region) in regions.iter() {
-                let text = &*self.preprocess(region, &mut cursor_total);
-                let text_trimmed = text.trim_end_matches(|c| c == '\r' || c == '\n');
-                write!(
-                    handle,
-                    "{}",
-                    as_terminal_escaped(
-                        style,
-                        text_trimmed,
-                        true_color,
-                        colored_output,
-                        italics,
-                        background_color
-                    )
-                )?;
+                let ansi_iterator = AnsiCodeIterator::new(region);
+                for chunk in ansi_iterator {
+                    match chunk {
+                        // ANSI escape passthrough.
+                        (ansi, true) => {
+                            self.ansi_style.update(ansi);
+                            write!(handle, "{}", ansi)?;
+                        }
 
-                if text.len() != text_trimmed.len() {
-                    if let Some(background_color) = background_color {
-                        let mut ansi_style = Style::default();
-                        ansi_style.background = to_ansi_color(background_color, true_color);
-                        let width = if cursor_total <= cursor_max {
-                            cursor_max - cursor_total + 1
-                        } else {
-                            0
-                        };
-                        write!(handle, "{}", ansi_style.paint(" ".repeat(width)))?;
+                        // Regular text.
+                        (text, false) => {
+                            let text = &*self.preprocess(text, &mut cursor_total);
+                            let text_trimmed = text.trim_end_matches(|c| c == '\r' || c == '\n');
+
+                            write!(
+                                handle,
+                                "{}",
+                                as_terminal_escaped(
+                                    style,
+                                    &format!("{}{}", self.ansi_style, text_trimmed),
+                                    true_color,
+                                    colored_output,
+                                    italics,
+                                    background_color
+                                )
+                            )?;
+
+                            if text.len() != text_trimmed.len() {
+                                if let Some(background_color) = background_color {
+                                    let mut ansi_style = Style::default();
+                                    ansi_style.background =
+                                        to_ansi_color(background_color, true_color);
+                                    let width = if cursor_total <= cursor_max {
+                                        cursor_max - cursor_total + 1
```

Comment:
Thanks for working on this issue :+1: 

Thought I'd try to assist in making progress on this PR. I currently find the code a bit hard to follow (but I must admit I did not yet sit down and studied it in detail). Some thoughts on things that could perhaps be done to make the code easier to code review, understand and maintain:

* This line is currently indented 10 levels. Would it be possible to introduce helper functions for improved code structure and thus code understandability? Possibly as separate, prepatory and self-contained PRs beforehand.
* "tricky" code like this is very helpful to have automated tests for. I noticed you added some tests. Do they cover all relevant code? I'm currently experimenting with code coverage, and hope to be able to answer that question myself at some point, but thought I'd ask the question here anyway.

Reply:
I actually don't really remember what this is for. It looks like it highlights until the end of the line for `--wrap=never` mode?

Reply:
Yes. It computes the amount of spaces that need to be added in order to extend the line-highlighting to the right boder of the terminal.

Reply:
This line is not changed when ignoring whitespace changes while diffing. So I'll just go ahead and resolve this conversation.